packages/office-ui-fabric-react/src/components/FocusZone/FocusZone.tsx

```diff
@@ -201,37 +230,120 @@ export class FocusZone extends BaseComponent<IFocusZoneProps, {}> implements IFo
+
+    if (doc) {
+      const focusedElement = doc.activeElement as HTMLElement;
+
+      // Only update the index path if we are not parked on the root.
+      if (focusedElement !== root) {
+        const shouldRestoreFocus = elementContains(root, focusedElement);
+
+        this._lastIndexPath = shouldRestoreFocus ? this._getIndexPath(doc.activeElement as HTMLElement) : undefined;
+      }
+    }
+  }
+
   private _onFocus = (ev: React.FocusEvent<HTMLElement>): void => {
     const { onActiveElementChanged, doNotAllowFocusEventToPropagate, onFocusNotification } = this.props;
+    const isImmediateDescendant = this._isImmediateDescendantOfZone(ev.target as HTMLElement);
+    let newActiveElement: HTMLElement | undefined;
 
     if (onFocusNotification) {
       onFocusNotification();
     }
 
-    if (this._isImmediateDescendantOfZone(ev.target as HTMLElement)) {
-      this._activeElement = ev.target as HTMLElement;
-      this._setFocusAlignment(this._activeElement);
+    if (isImmediateDescendant) {
+      newActiveElement = ev.target as HTMLElement;
     } else {
       let parentElement = ev.target as HTMLElement;
 
       while (parentElement && parentElement !== this._root.current) {
         if (isElementTabbable(parentElement) && this._isImmediateDescendantOfZone(parentElement)) {
-          this._activeElement = parentElement;
+          newActiveElement = parentElement;
           break;
         }
         parentElement = getParent(parentElement, ALLOW_VIRTUAL_ELEMENTS) as HTMLElement;
       }
     }
 
-    if (onActiveElementChanged) {
-      onActiveElementChanged(this._activeElement as HTMLElement, ev);
+    if (newActiveElement && newActiveElement !== this._activeElement) {
+      this._activeElement = newActiveElement;
+
+      if (isImmediateDescendant) {
+        this._setFocusAlignment(this._activeElement);
+      }
+
+      if (onActiveElementChanged) {
+        onActiveElementChanged(this._activeElement as HTMLElement, ev);
+      }
     }
 
     if (doNotAllowFocusEventToPropagate) {
       ev.stopPropagation();
     }
   };
 
+  private _setParkedFocus(isParked: boolean): void {
+    const { current: root } = this._root;
+    if (root) {
+      if (isParked) {
+        if (!this.props.allowFocusRoot) {
+          this._parkedTabIndex = root.getAttribute('tabindex');
+          root.setAttribute('tabindex', '-1');
+        }
+        root.focus();
+      } else {
+        if (!this.props.allowFocusRoot) {
+          if (this._parkedTabIndex !== null) {
+            root.setAttribute('tabindex', this._parkedTabIndex);
+          } else {
+            root.removeAttribute('tabindex');
+          }
+        }
+      }
+    }
+  }
+
+  private _getElementByIndexPath(path: number[]): HTMLElement | undefined {
+    let element = this._root.current as HTMLElement;
+
+    for (const index of path) {
+      const nextChild = element.children[Math.min(index, element.children.length - 1)] as HTMLElement;
```

Comment:
Isn't it possible for this to lead to a random element getting focus? Lets say the root has 6 children, and focus is on a child of index 3. Index 3's entire element is removed. Root still has 5 children, so there is still an index 3 and it will follow the path down to another element. In the case of a list that would probably be okay but in other cases it could lead to focus getting put on seemingly unrelated things.